lib/Wikia/src/Service/User/Permissions/PermissionsServiceImpl.php

```diff
@@ -0,0 +1,391 @@
+		) );
+	}
+
+	public function getExplicitLocalGroups(\User $user ) {
+		$this->loadLocalGroups( $user->getId() );
+		return $this->localExplicitUserGroups[ $user->getId() ];
+	}
+
+	public function getExplicitGlobalGroups(\User $user ) {
+		$this->loadGlobalUserGroups( $user->getId() );
+		return $this->globalExplicitUserGroups[ $user->getId() ];
+	}
+
+	public function isInGroup(\User $user, $group ) {
+		return in_array( $group, $this->getEffectiveGroups( $user ) );
+	}
+
+	/**
+	 * Return memcache key used for storing groups for a given user
+	 *
+	 * @param $userId
+	 * @return string memcache key
+	 */
+	static public function getMemcKey( $userId ) {
+		return wfSharedMemcKey( __CLASS__, 'permissions-groups', $userId );
+	}
+
+	/**
+	 * Get the list of implicit group memberships this user has.
+	 * This includes 'user' if logged in, '*' for all accounts,
+	 * and autopromoted groups
+	 * @param \User $user
+	 * @param $reCacheAutomaticGroups
+	 * @return string[] internal group names
+	 */
+	public function getAutomaticGroups(\User $user, $reCacheAutomaticGroups = false ) {
+		if ( !$reCacheAutomaticGroups && isset( $this->implicitUserGroups[ $user->getId() ] ) ) {
+			$implicitGroups = $this->implicitUserGroups[ $user->getId() ];
+		} else {
+			$implicitGroups = array( '*' );
+			if ( $user->getId() ) {
+				$implicitGroups[] = 'user';
+
+				$implicitGroups = array_unique( array_merge(
+					$implicitGroups,
+					\Autopromote::getAutopromoteGroups( $user )
+				) );
+				$this->implicitUserGroups[ $user->getId() ] = $implicitGroups;
+			}
+		}
+
+		return $implicitGroups;
+	}
+
+	/**
+	 * Get the list of explicit and implicit group memberships this user has.
+	 * This includes all explicit groups, plus 'user' if logged in,
+	 * '*' for all accounts, and autopromoted groups
+	 * @param \User $user
+	 * @param $reCacheAutomaticGroups
+	 * @return string[] internal group names
+	 */
+	public function getEffectiveGroups(\User $user, $reCacheAutomaticGroups = false ) {
+
+		return array_unique( array_merge(
+			$this->getExplicitGroups( $user ),
+			$this->getAutomaticGroups( $user, $reCacheAutomaticGroups )
+		) );
+	}
+
+	/**
+	 * Get the permissions this user has.
+	 * @param \User $user
+	 * @return string[] permission names
+	 */
+	public function getPermissions(\User $user ) {
+		$userId = $user->getId();
+
+		if ( isset( $this->userPermissions[ $userId ] ) ) {
+			$permissions = $this->userPermissions[ $userId ];
+		} else {
+			$permissions = $this->permissionsConfiguration->getGroupPermissions( $this->getEffectiveGroups( $user ) );
+			wfRunHooks( 'UserGetRights', array( $user, &$permissions ) );
+			if ( !empty( $userId ) ) {
+				$this->userPermissions[ $userId ] = array_values( $permissions );
+			}
+		}
+		return $permissions;
+	}
+
+	private function loadLocalGroups( $userId ) {
+		if ( !empty( $userId ) && !isset( $this->localExplicitUserGroups[ $userId ] ) ) {
+			$dbr = wfGetDB( DB_MASTER );
```

Comment:
Possibly. One thing I would keep in mind though is that this code will be deleted in a few weeks and replaced with an invocation of the user permissions service. So I would suggest to leave this implementation as is (as it's tested and working like that for years) and then we I make step 2 of moving it to the service, think about an appropriate caching strategy.